pkg/grpc/ratelimit/ratelimit_impl.go

```diff
@@ -13,27 +14,37 @@ import (
 )
 
 type rateLimiter struct {
-	mutex              sync.Mutex
-	tokenBucketLimiter *rate.Limiter
+	mutex sync.Mutex
+
+	maxThrottleDuration time.Duration
+	tokenBucketLimiter  *rate.Limiter
 }
 
-func newRateLimiter(maxPerSec int) *rateLimiter {
+func newRateLimiter(maxPerSec int, maxThrottleDuration time.Duration) *rateLimiter {
 	limit := rate.Inf
 	if maxPerSec > 0 {
 		limit = rate.Every(time.Second / time.Duration(maxPerSec))
 	}
 
 	// When no limit is set, we use "rate.Inf," and burst is disregarded.
 	limiter := &rateLimiter{
-		tokenBucketLimiter: rate.NewLimiter(limit, maxPerSec),
+		maxThrottleDuration: maxThrottleDuration,
+		tokenBucketLimiter:  rate.NewLimiter(limit, maxPerSec),
 	}
 
 	return limiter
 }
 
 // Limit implements "ratelimit.Limiter" interface.
 func (limiter *rateLimiter) Limit() bool {
-	return !limiter.tokenBucketLimiter.Allow()
+	if limiter.maxThrottleDuration < time.Second {
+		return !limiter.tokenBucketLimiter.Allow()
+	}
+
+	ctx, cancelFnc := context.WithDeadline(context.Background(), time.Now().Add(limiter.maxThrottleDuration))
```

Comment:
What's the difference between using withdeadline and withtimeout

Reply:
Good catch. There is no difference 
`WithTimeout returns WithDeadline(parent, time.Now().Add(timeout)).`

```suggestion
	ctx, cancelFnc := context.WithTimeout(context.Background(), limiter.maxThrottleDuration)
```